#' Create an empty pedigree
#'
#' \code{create_pedFile} initializes an empty data frame with all of the fields required by \code{sim_ped} to simulate a pedigree.
#'
#' @return An empty data frame with all fields required by \code{sim_ped} to generate a pedigree.
#' @keywords internal
#'
create_pedFile = function(){
  data.frame(FamID = numeric(),
             ID = numeric(),
             sex = numeric(),
             dadID = numeric(),
             momID = numeric(),
             affected = logical(),
             DA1 = numeric(),
             DA2 = numeric(),
             birthYr = numeric(),
             onsetYr = numeric(),
             deathYr = numeric(),
             RR = numeric(),
             available = logical(),
             Gen = numeric(),
             do_sim = logical(),
             stringsAsFactors = FALSE)
}

#' Create a new seed founder
#'
#' Create new seed founder for pedigree.
#'
#' @inheritParams sim_RVped
#'
#' @return \code{new_founder_info} the pedigree information for the new mate.
#' @keywords internal
create_founder = function(FamID, GRR, carrier_prob,
                          RVfounder, founder_byears){

  founder_dat <- sim_founderRVstatus(GRR, carrier_prob,
                                     RVfounder)

  new_founder_info <- data.frame(FamID = FamID,
                                 ID = 1,
                                 sex = round(runif(1)),
                                 dadID = NA,
                                 momID = NA,
                                 affected = F,
                                 DA1 = founder_dat[[1]][1],
                                 DA2 = founder_dat[[1]][2],
                                 birthYr = round(runif(1,
                                                       min = founder_byears[1],
                                                       max = founder_byears[2])),
                                 onsetYr = NA,
                                 deathYr = NA,
                                 RR = founder_dat[[2]],
                                 available = T,
                                 Gen = 1,
                                 do_sim = T,
                                 stringsAsFactors = FALSE)

  return(new_founder_info)
}

#' Create a new mate
#'
#' Create new mating partner in pedigree.
#'
#' @inheritParams sim_RVped
#' @param partner_info  Data.frame with 1 row. All pedigree information for the individual requiring a new mate, must contain all fields generated by \code{\link{create_pedFile}}.
#' @param last_id Numeric.  The last id used in the pedigree.
#'
#' @return \code{new_mate_info} the pedigree information for the new mate.
#' @return \code{last_id} the updated value of last_id after adding the new mate.
#'
#' @keywords internal
#'
#'
create_mate = function(partner_info, last_id,
                       GRR, carrier_prob,
                       RVfounder){

  new_mate_info <- data.frame(FamID = partner_info$FamID,
                              ID = last_id + 1,
                              sex = abs(partner_info$sex - 1),
                              dadID = NA,
                              momID = NA,
                              affected = F,
                              DA1 = 0,
                              DA2 = 0,
                              birthYr = NA,
                              onsetYr = NA,
                              deathYr = NA,
                              RR = 1,
                              available = F,
                              Gen = partner_info$Gen,
                              do_sim = F,
                              stringsAsFactors = FALSE)

  mate_return <- list(new_mate_info, last_id + 1)
  return(mate_return)
}

#' Create a new offspring
#'
#' Create new offspring in pedigree.
#'
#' @inheritParams create_mate
#' @inheritParams sim_RVped
#' @param dad_info Data.frame with 1 row. All pedigree information of offspring's father, must contain all fields generated by \code{\link{create_pedFile}}.
#' @param mom_info Data.frame with 1 row. All pedigree information of offspring's mother, must contain all fields generated by \code{\link{create_pedFile}}.
#' @param byear The birth year of offspring.
#'
#' @return new_child_info the pedigree information for the new child
#' @return last_id the updated value of last_id after adding new offspring
#'
#' @keywords internal
#'
#' @importFrom stats runif
#'
create_offspring = function(dad_info, mom_info, byear, last_id, GRR){
  new_child_info <- data.frame(FamID = dad_info$FamID,
                               ID = last_id + 1,
                               sex = round(runif(1)),
                               dadID = dad_info$ID,
                               momID = mom_info$ID,
                               affected = F,
                               DA1 = sample(x = c(dad_info$DA1, dad_info$DA2), size = 1),
                               DA2 = sample(x = c(mom_info$DA1, mom_info$DA2), size = 1),
                               birthYr = byear,
                               onsetYr = NA,
                               deathYr = NA,
                               RR = NA,
                               available = T,
                               Gen = dad_info$Gen + 1,
                               do_sim = T,
                               stringsAsFactors = FALSE)
  new_child_info$RR <- ifelse(sum(new_child_info$DA1, new_child_info$DA2) == 0,
                              1, GRR)
  child_return <- list(new_child_info, last_id + 1)
  return(child_return)
}


#' Simulate a nuclear family from a single founder
#'
#' Simulate all life events for an individual and create a nuclear family, when appropriate.
#'
#' @param found_info Data.frame with 1 row. All pedigree information for the founder, must contain all fields generated by \code{\link{create_pedFile}}.
#' @inheritParams create_mate
#' @inheritParams sim_RVped
#'
#' @keywords internal
#' @return A pedigree with updated life events for the founder, and
#' additional information for mate and offspring, when offspring are generated.
#'
sim_nFam = function(found_info, stop_year, last_id,
                    hazard_rates, birth_range, NB_params,
                    GRR, carrier_prob, RVfounder, intro_RV){

  nfam_ped <- found_info

  #Simulate life steps for founder
  sim_years <- sim_life(hazard_rates, GRR, carrier_prob,
                        RV_status = any(found_info[, c(7:8)] == 1),
                        YOB = found_info$birthYr, stop_year,
                        birth_range, NB_params)



  # update disease status and onset year if onset occured prior to stop_year
  if (is.element("Onset", colnames(sim_years))) {
    nfam_ped$affected <- T
    nfam_ped$onsetYr <- sim_years[colnames(sim_years) == "Onset"]
  } else {
    nfam_ped$affected <- F
  }

  # update death status and death year if onset occured prior to stop_year
  if (is.element("Death", colnames(sim_years))) {
    nfam_ped$deathYr <- sim_years[colnames(sim_years) == "Death"]
  }

  #store the birth years of each child
  birth_events <- sim_years[colnames(sim_years) == "Child"]

  if (length(birth_events) > 0) {
    #add a mate
    new_mate <- create_mate(partner_info = nfam_ped[1,], last_id,
                            GRR, carrier_prob,
                            RVfounder)

    nfam_ped <- rbind(nfam_ped, new_mate[[1]])
    last_id <- new_mate[[2]]

    #store info for mom and dad
    dad <- nfam_ped[which(nfam_ped$sex == 0), ]
    mom <- nfam_ped[which(nfam_ped$sex == 1), ]

    for (k in 1:length(birth_events)) {
      #add child
      new_child <- create_offspring(dad_info = dad, mom_info = mom,
                                    byear = birth_events[k], last_id, GRR)
      nfam_ped <- rbind(nfam_ped, new_child[[1]])
      last_id <- new_child[[2]]
    }
  }

  #set do_sim to FALSE for individual whose life events we just simulated
  nfam_ped$do_sim[1] = F

  sim_fam_return = list(nfam_ped, last_id)
  return(sim_fam_return)
}

#' Simulate a pedigree
#'
#' Please note the distinction between \code{sim_ped} and \code{sim_RVped}.  Pedigrees simulated using \code{sim_ped} do not account for study design.  To simulate a pedigree ascertained to contain multiple family members affected by a disease please use \code{\link{sim_RVped}}.
#'
#'
#' To introduce the rare variant to the pedigree, We allow users to choose from one of the following two assumptions:
#' \enumerate{
#' \item Assume that the variant is rare enough that a single copy has been introduced by one founder, and begin the simulation of the pedigree with this founder, as in Bureau (2014).
#' \item Simulate the starting founder's rare-variant status with probability equal to the carrier probability of the rare variant in the population.  We note that under this setting pedigrees may not segregate the rare variant.
#' }
#' The \code{sim_ped} function starts simulating the pedigree by generating the birth year for the starting founder, uniformly between the years specified by \code{founder_byears}.  Next, all life events are simulated for the founder via \code{\link{sim_life}}.  Possible life events include: reproduction, disease onset, and death.  We only allow disease onset to occur once, i.e. no remission.  Computationally, this implies that after disease onset, the waiting time to death is always simulated using the age-specific mortality rates for the \emph{affected} population.  Life events for individuals who have inherited the rare variant are simulated such that their relative-risk of disease is \code{GRR}, according to a proportional hazards model.  The relative-risk of disease onset for individuals who have not inherited the causal variant is assumed to be 1.  Any life events that occur after \code{stop_year} are censored.
#'
#' When segregating in the pedigree, the rare variant is transmitted from parent to offspring according to Mendel's laws.  The process of simulating life events is repeated for any offspring that are produced before \code{stop_year}.
#'
#'
#' @inheritParams sim_RVped
#'
#' @return The simulated pedigree.
#' @export
#' @importFrom stats runif
#'
#' @references OUR MANUSCRIPT
#' @references Ken-Ichi Kojima, Therese M. Kelleher. (1962), \emph{Survival of Mutant Genes}. The American Naturalist 96, 329-346.
#' @references Alexandre Bureau, Samuel G. Younkin, Margaret M. Parker, Joan E. Bailey-Wilson, Mary L. Marazita, Jeffrey C. Murray, Elisabeth Mangold, Hasan Albacha-Hejazi, Terri H. Beaty, and Ingo Ruczinski (2014). \emph{Inferring rare disease risk variants based on exact probabilities of sharing by multiple affected relatives.} Bioinformatics; Vol. 30, No. 15, pp. 2189-2196.
#'
#' @section See Also:
#' \code{\link{sim_RVped}}
#'
#' @examples
#' data(AgeSpecific_Hazards)
#'
#' #Simulate a random pedigree
#' set.seed(22)
#' ex_ped <- sim_ped(hazard_rates = hazard(hazardDF = AgeSpecific_Hazards),
#'                   GRR = 10,
#'                   FamID = 1,
#'                   founder_byears = c(1900, 1910),
#'                   stop_year = 2015)
#'
#' ex_ped
#' plot(ex_ped, legendLocation = "bottomleft")
#' plot(ex_ped, ref_year = 2015,
#'      cex= 0.75, symbolsize = 1.25,
#'      legendLocation = "bottomleft")
#'
#' summary(ex_ped)
#'
#' #Simulate a random pedigree
#' set.seed(22)
#' ex_ped <- sim_ped(hazard_rates = hazard(hazardDF = AgeSpecific_Hazards),
#'                   RVfounder = TRUE,
#'                   GRR = 10,
#'                   FamID = 1,
#'                   founder_byears = c(1900, 1910),
#'                   stop_year = 2015)
#'
#' ex_ped
#' plot(ex_ped)
#' plot(ex_ped, ref_year = 2015,
#'      cex= 0.75, symbolsize = 1.25,
#'      legendLocation = "topleft")
#' summary(ex_ped)
#'
sim_ped = function(hazard_rates, GRR,
                   FamID, founder_byears, stop_year,
                   carrier_prob = 0.002,
                   RVfounder = FALSE,
                   birth_range = c(18, 45),
                   NB_params = c(2, 4/7)){

  if(!(RVfounder %in% c(T, F))){
    stop ('Please set RVfounder to TRUE or FALSE.')
  }

  if (carrier_prob <= 0 | carrier_prob >= 1){
    stop ('carrier_prob must be a value between 0 and 1')
  } else if (carrier_prob > 0.002) {
    warning('carrier_prob > 0.002: sim_RVped is intended for simulating the transmission of rare variants.')
  }

  if(missing(stop_year)){
    stop_year <- as.numeric(format(Sys.Date(),'%Y'))
  }

  fam_ped <- create_pedFile()
  fam_ped[1, ] <- create_founder(FamID, GRR, carrier_prob,
                                 RVfounder, founder_byears)

  last_id <- 1
  last_gen <- 1

  #store the ID of individuals for whom we need to simulate life events
  re_sim <- fam_ped$ID[which(fam_ped$do_sim & fam_ped$Gen == last_gen)]
  while (length(re_sim) > 0) {
    for (k in 1:length(re_sim)) {
      newKin <- sim_nFam(found_info = fam_ped[which(fam_ped$ID == re_sim[k]),],
                         stop_year, last_id,
                         hazard_rates, birth_range, NB_params,
                         GRR, carrier_prob,
                         RVfounder)

      #replace individual by their simulated self and add family members when necessary
      fam_ped <- rbind(fam_ped[-which(fam_ped$ID == re_sim[k]), ],
                       newKin[[1]])

      last_id <- newKin[[2]]
    }
    last_gen <- max(fam_ped$Gen)
    re_sim <- fam_ped$ID[which(fam_ped$do_sim & fam_ped$Gen == last_gen)]
  }

  rownames(fam_ped) <- NULL

  return(ped(fam_ped[, c(1:14)]))
}


#' Simulate a pedigree ascertained to contain multiple disease-affected relatives
#'
#' \code{sim_RVped} simulates a pedigree ascertained to contain multiple affected members, selects a proband, and trims the pedigree to contain only those individuals that are recalled by the proband.
#'
#' When \code{RV_founder = TRUE}, all simulated pedigrees will segregate a genetic susceptibility variant.  In this scenario, we assume that the variant is rare enough that it has been introduced by one founder, and we begin the simulation of the pedigree with this founder.  Alternatively, when \code{RV_founder = FALSE} we simulate the starting founder's causal variant status with probability \code{carrier_prob}.  When \code{RV_founder = FALSE} pedigrees may not segregate the genetic susceptibility variant.  The default selection is \code{RV_founder = FALSE}.  Additionally, we note that \code{sim_RVpedigree} is intended for rare causal variants; users will recieve a warning if \code{carrier_prob > 0.002}.
#'
#' We note that when \code{GRR = 1}, pedigrees do not segregate the causal variant regardless of the setting selected for \code{RVfounder}.  When the causal variant is introduced to the pedigree we transmit it from parent to offspring according to Mendel's laws.
#'
#' We begin simulating the pedigree by generating the year of birth, uniformly, between the years specified in \code{founder_byears} for the starting founder.  Next, we simulate this founder's life events using the \code{\link{sim_life}} function, and censor any events that occur after the study \code{stop_year}.  Possible life events include: reproduction, disease onset, and death. We continue simulating life events for any offspring, censoring events which occur after the study stop year, until the simulation process terminates.  We do not simulate life events for marry-ins, i.e. individuals who mate with either the starting founder or offspring of the starting founder.
#'
#' We do not model disease remission. Rather, we impose the restriction that individuals may only experience disease onset once, and remain affected from that point on.  If disease onset occurs then we apply the hazard rate for death in the affected population.
#'
#' \code{sim_RVped} will only return ascertained pedigrees with at least \code{num_affected} affected individuals.  That is, if a simulated pedigree does not contain at least \code{num_affected} affected individuals \code{sim_RVped} will discard the pedigree and simulate another until the condition is met.  We note that even for \code{num_affected = 2}, \code{sim_RVped} can be computationally expensive.  To simulate a pedigree with no proband, and without a minimum number of affected members use instead \code{\link{sim_ped}}.
#'
#' Upon simulating a pedigree with \code{num_affected} individuals, \code{sim_RVped} chooses a proband from the set of available candidates.  Candidates for proband selection must have the following qualities:
#' \enumerate{
#'   \item experienced disease onset between the years specified by \code{ascertain_span},
#'   \item if less than \code{num_affected} - 1 individuals experienced disease onset prior to the lower bound of \code{ascertain_span}, a proband is chosen from the affected individuals, such that there were at least \code{num_affected} affected individuals when the pedigree was ascertained through the proband.
#' }
#'
#' After the proband is selected, the pedigree is trimmed based on the proband's recall probability of his or her relatives.  This option is included to allow researchers to model the possibility that a proband either cannot provide a complete family history or that they explicitly request that certain family members not be contacted.  If \code{recall_probs} is missing, the default values of four times the kinship coefficient, as defined by Thompson (see references), between the proband and his or her relatives are assumed.  This has the effect of retaining all first degree relatives with probability 1, retaining all second degree relatives with probability 0.5, retaining all third degree relatives with probability 0.25, etc.  Alternatively, the user may specify a list of length \eqn{l}, such that the first \eqn{l-1} items represent the respective recall probabilities for relatives of degree \eqn{1, 2, ... , l-1} and the \eqn{l^{th}} item represents the recall probability of a relative of degree \eqn{l} or greater. For example, if \code{recall_probs = c(1, 0.75, 0.5)}, then all first degree relatives (i.e. parents, siblings, and offspring) are retained with probability 1, all second degree relatives (i.e. grandparents, grandchildren, aunts, uncles, nieces and nephews) are retained with probability 0.75, and all other relatives are retained with probability 0.5. To simulate fully ascertained pedigrees, simply specify \code{recall_probs = c(1)}.
#'
#' In the event that a trimmed pedigree fails the \code{num_affected} condition,  \code{sim_RVped} will discard that pedigree and simulate another until the condition is met.  For this reason, the values specified for \code{recall_probs} affect computation time.
#'
#' @param hazard_rates An object of class \code{hazard}, created by \code{\link{hazard}}.
#' @param GRR Numeric. The genetic relative-risk of disease, i.e. the relative-risk of disease for individuals who carry at least one copy of the causal variant.
#' @param carrier_prob  Numeric.  The carrier probability for all causal variants with relative-risk of disease \code{GRR}.  By default, \code{carrier_prob}\code{ = 0.002}
#' @param RVfounder Logical.  Indicates if all pedigrees segregate the rare, causal variant.  By default, \code{RVfounder = FALSE} See details.
#' @param founder_byears Numeric vector of length 2.  The span of years from which to simulate, uniformly, the birth year for the founder who introduced the rare variant to the pedigree.
#' @param ascertain_span Numeric vector of length 2.  The year span of the ascertainment period.  This period represents the range of years during which the proband developed disease and the family would have been ascertained for multiple affected relatives.
#' @param num_affected Numeric.  The minimum number of affected individuals in the pedigree.
#' @param FamID Numeric. The family ID to assign to the simulated pedigree.
#' @param recall_probs Numeric. The proband's recall probabilities for relatives, see details.  If missing, four times kinship coefficient between the proband and the relative is used.
#' @param stop_year Numeric. The last year of study.  If missing, the current year is used.
#' @param birth_range Numeric vector of length 2. The minimum and maximum allowable ages, in years, between which individuals may reproduce.  By default, \code{birth_range}\code{ = c(18, 45)}.
#' @param NB_params Numeric vector of length 2. The size and probability parameters of the negative binomial distribution used to model the number of children per household.  By default, \code{NB_params}\code{ = c(2, 4/7)}, due to the investigation of Kojima and Kelleher (1962).
#'
#' @return  A list containing the following data frames:
#' @return \item{\code{full_ped} }{The full pedigree, prior to proband selection and trimming.}
#' @return \item{\code{ascertained_ped} }{The ascertained pedigree, with proband selected and trimmed according to proband recall probability.  See details.}
#' @export
#'
#' @references OUR MANUSCRIPT
#' @references Ken-Ichi Kojima, Therese M. Kelleher. (1962), \emph{Survival of Mutant Genes}. The American Naturalist 96, 329-346.
#' @references Thompson, E. (2000). \emph{Statistical Inference from Genetic Data on Pedigrees.} NSF-CBMS Regional Conference Series in Probability and Statistics, 6, I-169. Retrieved from http://www.jstor.org.proxy.lib.sfu.ca/stable/4153187
#'
#'
#' @section See Also:
#' \code{\link{sim_ped}}, \code{\link{trim.ped}}, \code{\link{sim_life}}
#'
#' @examples
#' #Read in age-specific hazards
#' data(AgeSpecific_Hazards)
#'
#' #Simulate pedigree ascertained for multiple affected individuals
#' set.seed(2410)
#' ex_RVped <- sim_RVped(hazard_rates = hazard(hazardDF = AgeSpecific_Hazards),
#'                      GRR = 10,
#'                      RVfounder = TRUE,
#'                      FamID = 1,
#'                      founder_byears = c(1900, 1920),
#'                      ascertain_span = c(1995, 2015),
#'                      num_affected = 2,
#'                      stop_year = 2017,
#'                      recall_probs = c(1, 0.5))
#'
#'
#' # Original pedigree prior to proband selection and trimming
#' summary(ex_RVped[[1]])
#' plot(ex_RVped[[1]])
#'
#' # The ascertained pedigree
#' summary(ex_RVped[[2]])
#' plot(ex_RVped[[2]])
#'
#'
#' # Simulate ascertained pedigree but this time set RVfounder = F.
#' # Under this setting pedigrees segregate a causal
#' # variant with probability equal to carrier prob.
#'
#'
sim_RVped = function(hazard_rates, GRR,
                    num_affected, ascertain_span,
                    FamID, founder_byears,
                    stop_year, recall_probs,
                    carrier_prob = 0.002,
                    RVfounder = FALSE,
                    birth_range = c(18, 45),
                    NB_params = c(2, 4/7)){

  if(!(RVfounder %in% c(FALSE, TRUE))){
    stop ('Please set RVfounder to TRUE or FALSE.')
  }

  if (length(ascertain_span) != 2 | ascertain_span[1] >= ascertain_span[2]){
    stop ('please provide appropriate values for ascertain_span')
  }


  if (num_affected <= 0){
    stop ('num_affected < 1: To simulate pedigrees that do not consider the number of disease-affected relatives please use sim_ped.')
  }

  if (GRR > 0 & GRR < 1){
    warning('Setting GRR < 1 can significantly increase computation time')
  }

  if(!missing(recall_probs)) {
    if (any(recall_probs > 1) | any(recall_probs < 0) ){
      stop ('recall probabilities must be between 0 and 1')
    } else if (any(recall_probs != cummin(recall_probs))){
      warning('Nondecreasing values specified for recall_probs')
    } else if (recall_probs[1] != 1){
      warning('recall_probs: First-degree relatives may not be recalled.')
    }
  }

  if(missing(stop_year)){
    stop_year <- as.numeric(format(Sys.Date(),'%Y'))
  }

  ascertained <- FALSE
  while(ascertained == FALSE){
    #generate pedigree
    fam_ped <- sim_ped(hazard_rates, GRR, FamID,
                      founder_byears, stop_year, carrier_prob,
                      RVfounder, birth_range, NB_params)

    #check to see if pedigree is ascertained
    check_pedigree <- is_ascertained(ped_file = fam_ped, num_affected,
                                     ascertain_span, recall_probs)

    #store updated pedigree
    ascertained <- check_pedigree[[1]]
    ascertained_ped <- check_pedigree[[2]]
  }

  #return original and trimmed pedigrees
  return(list(full_ped = fam_ped, ascertained_ped = ascertained_ped))
}
